### ISSUE TITLE
OnDestroy override issue - Angular 9+

### ISSUE THREAD

**Issue by AvshT** [angular/angular#38043]
# 🐞 bug report

### Affected Package

 The issue is caused by package **@angular/core**

### Is this a regression?

 Yes, the previous version in which this bug was not present was: **Angular 8**

### Description

 
Consider a @Component what implements OnDestroy,
on the **constructor** there is a case I must set the following code:

**this.OnDestroy = ()=>{
         alert('Invoke');
}**

On Angular 8, **it use to work**, but somehow in Angular 9 (maybe its IVE issue) that not entering to the override func, instead it entering to the initial implemented func.

Can you explain why and if it really a bug?

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.901.6
@angular-devkit/core         9.1.6
@angular-devkit/schematics   9.1.6
@schematics/angular          9.1.6
@schematics/update           0.901.6
rxjs                         6.5.4
</code></pre>

**Comment by petebacondarwin** [comment]
This is kind of a duplicate of https://github.com/angular/angular/issues/30497, which is fixed by https://github.com/angular/angular/pull/35464. But in your case, you appear to be changing the `onDestroy()` method on the instance, whereas that PR will only allow you to change the method on the prototype.

I think you could workaround this right now by creating your class like:

```ts
@Component(...)
export class MyComponent implements NgOnDestroy {
  onDestroy() {
    this._onDestroy();
  }

  private _onDestroy() {}

  constructor() {
    this._onDestroy = () => {
      alert('Invoke');
    };
  }
}
```

**Comment by spock123** [comment]
@petebacondarwin there is no way to use arrow functions in general as lifecycle hooks, is there?

Is this because of some frsmework issues (inheritance) or is it a typescript issue? It's just for aesthetics, doing hooks as regular functions and all other methods as fat arrow functions just looks weird to my ocd brain ;)

**Comment by petebacondarwin** [comment]
@spock123 - I don't understand the question here. Could you give a concrete example of what is not working for you?

**Comment by spock123** [comment]
@petebacondarwin I'm sorry I did not intend to hijack this issue but I guess it's somewhat related:

If I have a component with the following code ngOnInit method:

`ngOnInit = () => console.log('Hello Init');`

- won't be called by the framework.

However this will be called:

`ngOnInit = function() { console.log('Hello Init'); }`


I was just wondering if it was somehow possible to use fat arrow functions when creating component lifecycle hooks

**Comment by petebacondarwin** [comment]
Sorry @spock123 - I am still not sure what the problem is here. The `ngOnInit()` method must be defined (even after #35464) before the app bootstraps. The following works for me:

```ts
@Component({selector: 'app-root', template: ''})
export class AppComponent {
  title = 'test-app';
}

(AppComponent.prototype as any).ngOnInit = () => {
  console.log('init!');
};
```

**Comment by petebacondarwin** [comment]
Using it directly in the component is effectively the same as doing it in the constructor. This has two problems:

1. It only changes the instance of the component and not the prototype, which Ivy will not consider. It only looks at the prototype.
2. It is too late for Ivy since it reads the prototype of the component at build time and takes a reference to the method.

**Comment by pkozlowski-opensource** [comment]
This currently works as designed - I've opened a new feature request with more info: #38241
